README.md
# viewd
render simple views based on request


```php
<?php

use Relay\RelayBuilder;
use Jnjxp\Viewd\Viewd;
use Aura\View\ViewFactory;
use Aura\Html\HelperLocatorFactory;
use Relay\Middleware\ExceptionHandler;
use Relay\Middleware\ResponseSender;
use Zend\Diactoros\Response;
use Zend\Diactoros\ServerRequestFactory;

require '../vendor/autoload.php';


$view = (new ViewFactory)->newInstance(
    (new HelperLocatorFactory)->newInstance()
);

$view->setLayout('default');

$templates = dirname(__DIR__) . '/templates';
$view->getViewRegistry()->setPaths(["{$templates}/views"]);
$view->getLayoutRegistry()->setPaths(["{$templates}/layouts"]);

$queue = [
    new ResponseSender(),
    new ExceptionHandler(new Response()),
    new Viewd($view)
];

$relay = (new RelayBuilder)->newInstance($queue);
$relay(ServerRequestFactory::fromGlobals(), new Response());


CHANGES.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [0.1.0] - 2016-01-28
### Changed
- Refactor to use `Responder` paradigm
- Change Versioning Paradigm
- Change Change Log paradigm

## [v1.0.0-alpha] - 2015-10-02
### Added
- Initial bad release with bad versioning

[0.1.0]: https://github.com/jnjxp/jnjxp.viewd/compare/v1.0.0-alpha...0.1.0
[v1.0.0-alpha]: https://github.com/jnjxp/jnjxp.viewd/releases/tag/v1.0.0-alpha


tests/ViewdTest.php
<?php
/**
* Jnjxp\Viewd
*
* PHP version 5
*
* This program is free software: you can redistribute it and/or modify it
* under the terms of the GNU Affero General Public License as published by
* the Free Software Foundation, either version 3 of the License, or (at your
* option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Affero General Public License for more details.
*
* You should have received a copy of the GNU Affero General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*
* @category  Tests
* @package   Jnjxp\Viewd
* @author    Jake Johns <jake@jakejohns.net>
* @copyright 2015 Jake Johns
* @license   http://www.gnu.org/licenses/agpl-3.0.txt AGPL V3
* @link      http://jakejohns.net
 */

namespace Jnjxp\Viewd;

use Aura\View\ViewFactory;
use Zend\Diactoros\ServerRequestFactory;
use Zend\Diactoros\Response;

/**
 * ViewdTest
 *
 * @category CategoryName
 * @package  PackageName
 * @author   Jake Johns <jake@jakejohns.net>
 * @license  http://www.gnu.org/licenses/agpl-3.0.txt AGPL V3
 * @link     http://jakejohns.net
 *
 */
class ViewdTest extends \PHPUnit_Framework_TestCase
{

    /**
     * setUp
     *
     * @return mixed
     *
     * @access public
     */
    public function setUp()
    {
        $this->view = (new ViewFactory)->newInstance();
        $this->viewd = new Viewd($this->view);

        $views = $this->view->getViewRegistry();

        $views->set(
            'viewd/index',
            function () {
                echo 'index content';
            }
        );

        $views->set(
            'viewd/foo',
            function () {
                echo 'foo content';
            }
        );

        $views->set(
            'error',
            function () {
                echo 'error content';
            }
        );

    }

    /**
     * testIndex
     *
     * @return mixed
     *
     * @access public
     */
    public function testIndex()
    {
        $response = new Response();
        $request = ServerRequestFactory::fromGlobals();
        $viewd = $this->viewd;

        $out = $viewd($request, $response);

        $this->assertEquals('index content', (string) $out->getBody());
        $this->assertEquals(200, $out->getStatusCode());
    }

    /**
     * testPath
     *
     * @return mixed
     *
     * @access public
     */
    public function testPath()
    {
        $response = new Response();
        $request = ServerRequestFactory::fromGlobals()
            ->withRequestTarget('/foo');

        $viewd = $this->viewd;

        $out = $viewd($request, $response);

        $this->assertEquals('foo content', (string) $out->getBody());
        $this->assertEquals(200, $out->getStatusCode());
    }

    /**
     * testAttr
     *
     * @return mixed
     *
     * @access public
     */
    public function testAttr()
    {
        $response = new Response();
        $request = ServerRequestFactory::fromGlobals()
            ->withAttribute('jnjxp/viewd:script', 'foo');

        $viewd = $this->viewd;

        $out = $viewd($request, $response);

        $this->assertEquals('foo content', (string) $out->getBody());
        $this->assertEquals(200, $out->getStatusCode());
    }

    /**
     * testError
     *
     * @return mixed
     *
     * @access public
     */
    public function testError()
    {
        $response = new Response();
        $request = ServerRequestFactory::fromGlobals()
            ->withRequestTarget('/foobarbaz');

        $viewd = $this->viewd;

        $out = $viewd($request, $response);

        $this->assertEquals('error content', (string) $out->getBody());
        $this->assertEquals(404, $out->getStatusCode());
    }

    /**
     * testNoError
     *
     * @return mixed
     *
     * @access public
     */
    public function testNoError()
    {
        $response = new Response();
        $response->getBody()->write('Other Body');
        $request = ServerRequestFactory::fromGlobals()
            ->withRequestTarget('/foobarbaz');

        $viewd = $this->viewd;

        $out = $viewd($request, $response);

        $this->assertEquals('Other Body', (string) $out->getBody());
        $this->assertEquals(200, $out->getStatusCode());
    }


    /**
     * testNoError
     *
     * @return mixed
     *
     * @access public
     */
    public function testStopError()
    {
        $response = new Response();
        $request = ServerRequestFactory::fromGlobals()
            ->withRequestTarget('/bingbamboom')
            ->withAttribute('jnjxp/viewd:no-error', true);

        $viewd = $this->viewd;

        $out = $viewd($request, $response);

        $this->assertEquals('', (string) $out->getBody());
        $this->assertEquals(200, $out->getStatusCode());
    }
}


src/SimpleResponder.php
<?php
/**
 * Jnjxp\Viewd
 *
 * PHP version 5
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @category  Responder
 * @package   Jnjxp\Viewd
 * @author    Jake Johns <jake@jakejohns.net>
 * @copyright 2016 Jake Johns
 * @license   http://www.gnu.org/licenses/agpl-3.0.txt AGPL V3
 * @link      http://jakejohns.net
 */


namespace Jnjxp\Viewd;

use Aura\View\View;

use Aura\Payload_Interface\PayloadInterface as Payload;
use Psr\Http\Message\ServerRequestInterface as Request;
use Psr\Http\Message\ResponseInterface as Response;

/**
 * StatusResponder
 *
 * @category Responser
 * @package  Jnjxp\Viewd
 * @author   Jake Johns <jake@jakejohns.net>
 * @license  http://www.gnu.org/licenses/agpl-3.0.txt AGPL V3
 * @link     http://jakejohns.net
 */
class SimpleResponder extends AbstractResponder
{

    /**
     * Attribute
     *
     * @var string
     *
     * @access protected
     */
    protected $attribute = 'view';

    /**
     * __construct
     *
     * @param View   $view      Aura\View
     * @param string $attribute request attribute of view name
     *
     * @access public
     */
    public function __construct(View $view, $attribute = 'view')
    {
        parent::__construct($view);
        $this->attribute = $attribute;
    }

    /**
     * Prepare
     *
     * @return void
     *
     * @access protected
     */
    protected function prepare()
    {
        parent::prepare();

        $name = $this->getViewName();
        $this->view->setView($name);
        $this->addPayloadData();
    }

    /**
     * Add payload data if payload is set
     *
     * @return void
     *
     * @access protected
     */
    protected function addPayloadData()
    {
        if ($this->payload) {
            $this->view->addData(
                [
                    'status' => $this->payload->getStatus(),
                    'output' => $this->payload->getOutput(),
                    'input' => $this->payload->getInput(),
                    'messages' => $this->payload->getMessages(),
                    'extras' => $this->payload->getExtras()
                ]
            );
        }
    }

    /**
     * Get view name from request attribute
     *
     * @return string
     *
     * @access protected
     */
    protected function getViewName()
    {
        if (! $name = $this->request->getAttribute($this->attribute)) {
            throw new \Exception('View attribute not set');
        }
        return $this->prefix . $name;
    }
}


src/StatusResponder.php
<?php
/**
 * Jnjxp\Viewd
 *
 * PHP version 5
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @category  Responder
 * @package   Jnjxp\Viewd
 * @author    Jake Johns <jake@jakejohns.net>
 * @copyright 2016 Jake Johns
 * @license   http://www.gnu.org/licenses/agpl-3.0.txt AGPL V3
 * @link      http://jakejohns.net
 */


namespace Jnjxp\Viewd;

use Aura\View\View;

use Aura\Payload_Interface\PayloadInterface as Payload;
use Psr\Http\Message\ServerRequestInterface as Request;
use Psr\Http\Message\ResponseInterface as Response;

/**
 * StatusResponder
 *
 * @category Responder
 * @package  Jnjxp\Viewd
 * @author   Jake Johns <jake@jakejohns.net>
 * @license  http://www.gnu.org/licenses/agpl-3.0.txt AGPL V3
 * @link     http://jakejohns.net
 */
class StatusResponder extends AbstractResponder
{

    /**
     * Prepare
     *
     * @return void
     *
     * @access protected
     */
    protected function prepare()
    {
        parent::prepare();

        $name = $this->getViewName();
        $this->view->setView($name);

        $method = $this->getMethodForPayload();
        $this->$method();
    }

    /**
     * GetMethodForPayload
     *
     * @return string
     *
     * @access protected
     */
    protected function getMethodForPayload()
    {
        if (! $this->payload) {
            return 'noContent';
        }

        $method = str_replace('_', '', strtolower($this->payload->getStatus()));
        return method_exists($this, $method) ? $method : 'unknown';
    }

    /**
     * GetViewName
     *
     * @return string
     *
     * @access protected
     */
    protected function getViewName()
    {
        if (! $this->payload) {
            return $this->prefix . 'no-content';
        }

        return $this->prefix
            . str_replace('-', '', strtolower($this->payload->getStatus()));
    }

    /**
     * Accepted
     *
     * @return void
     *
     * @access protected
     */
    protected function accepted()
    {
        $this->response = $this->response->withStatus(202);
        $this->view->addData($this->payload->getOutput());
    }

    /**
     * Created
     *
     * @return void
     *
     * @access protected
     */
    protected function created()
    {
        $this->response = $this->response->withStatus(201);
        $this->view->addData($this->payload->getOutput());
    }

    /**
     * Deleted
     *
     * @return void
     *
     * @access protected
     */
    protected function deleted()
    {
        $this->response = $this->response->withStatus(204);
        $this->view->addData($this->payload->getOutput());
    }

    /**
     * Error
     *
     * @return void
     *
     * @access protected
     */
    protected function error()
    {
        $this->response = $this->response->withStatus(500);
        $this->view->addData(
            [
                'input' => $this->payload->getInput(),
                'error' => $this->payload->getOutput(),
            ]
        );
    }

    /**
     * Failure
     *
     * @return void
     *
     * @access protected
     */
    protected function failure()
    {
        $this->response = $this->response->withStatus(400);
        $this->view->addData($this->payload->getInput());
    }

    /**
     * Found
     *
     * @return void
     *
     * @access protected
     */
    protected function found()
    {
        $this->response = $this->response->withStatus(200);
        $this->view->addData($this->payload->getOutput());
    }

    /**
     * NoContent
     *
     * @return void
     *
     * @access protected
     */
    protected function noContent()
    {
        $this->response = $this->response->withStatus(204);
    }

    /**
     * NotAuthenticated
     *
     * @return void
     *
     * @access protected
     */
    protected function notAuthenticated()
    {
        $this->response = $this->response->withStatus(401);
        $this->view->addData($this->payload->getInput());
    }

    /**
     * NotAuthorized
     *
     * @return void
     *
     * @access protected
     */
    protected function notAuthorized()
    {
        $this->response = $this->response->withStatus(403);
        $this->view->addData($this->payload->getInput());
    }

    /**
     * NotFound
     *
     * @return void
     *
     * @access protected
     */
    protected function notFound()
    {
        $this->response = $this->response->withStatus(404);
        $this->view->addData($this->payload->getInput());
    }

    /**
     * NotValid
     *
     * @return void
     *
     * @access protected
     */
    protected function notValid()
    {
        $this->response = $this->response->withStatus(422);
        $this->view->addData(
            [
            'input' => $this->payload->getInput(),
            'output' => $this->payload->getOutput(),
            'messages' => $this->payload->getMessages(),
            ]
        );
    }

    /**
     * Processing
     *
     * @return void
     *
     * @access protected
     */
    protected function processing()
    {
        $this->response = $this->response->withStatus(203);
        $this->view->addData($this->payload->getOutput());
    }

    /**
     * Success
     *
     * @return void
     *
     * @access protected
     */
    protected function success()
    {
        $this->response = $this->response->withStatus(200);
        $this->view->addData($this->payload->getOutput());
    }

    /**
     * Unknown
     *
     * @return void
     *
     * @access protected
     */
    protected function unknown()
    {
        $this->response = $this->response->withStatus(500);
        $this->view->addData(
            [
            'error' => 'Unknown domain payload status',
            'status' => $this->payload->getStatus(),
            ]
        );
    }

    /**
     * Updated
     *
     * @return void
     *
     * @access protected
     */
    protected function updated()
    {
        $this->response = $this->response->withStatus(303);
        $this->view->addData($this->payload->getOutput());
    }
}


src/ExceptionHandler.php
<?php
/**
 * Jnjxp\Viewd
 *
 * PHP version 5
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @category  Middleware
 * @package   Jnjxp\Viewd
 * @author    Jake Johns <jake@jakejohns.net>
 * @copyright 2016 Jake Johns
 * @license   http://www.gnu.org/licenses/agpl-3.0.txt AGPL V3
 * @link      http://jakejohns.net
 */


namespace Jnjxp\Viewd;

use Aura\View\View;

use Psr\Http\Message\ServerRequestInterface as Request;
use Psr\Http\Message\ResponseInterface as Response;

/**
 * ExceptionHandler
 *
 * @category Middleware
 * @package  Jnjxp\Viewd
 * @author   Jake Johns <jake@jakejohns.net>
 * @license  http://www.gnu.org/licenses/agpl-3.0.txt AGPL V3
 * @link     http://jakejohns.net
 */
class ExceptionHandler
{

    /**
     * Name of view to render
     *
     * @var string
     *
     * @access protected
     */
    protected $viewName = 'exception';

    /**
     * View
     *
     * @var View
     *
     * @access protected
     */
    protected $view;

    /**
     * Exception Response
     *
     * @var Response
     *
     * @access protected
     */
    protected $exceptionResponse;

    /**
     * __construct
     *
     * @param View     $view     Aura\View Object
     * @param Response $response Response for Exception 
     *
     * @return mixed
     *
     * @access public
     */
    public function __construct(View $view, Response $response)
    {
        $this->view = $view;
        $this->response = $response;
    }

    /**
     * __invoke
     *
     * @param Request  $request  PSR7 Request
     * @param Response $response PSR7 Response
     * @param callable $next     Next middleware chain
     *
     * @return Response
     *
     * @access public
     */
    public function __invoke(Request $request, Response $response, callable $next)
    {
        try {
            $response = $next($request, $response);
        } catch (Exception $e) {
            $response = $this->exceptionResponse->withStatus(500);
            $view = $this->view;
            $view->addData(
                [
                    'request' => $request,
                    'exception' => $e
                ]
            );
            $view->setView($this->viewName);
            $response->getBody()->write($view());
        }
        return $response;
    }
}


src/AbstractResponder.php
<?php
/**
 * Jnjxp\Viewd
 *
 * PHP version 5
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @category  Responder
 * @package   Jnjxp\Viewd
 * @author    Jake Johns <jake@jakejohns.net>
 * @copyright 2016 Jake Johns
 * @license   http://www.gnu.org/licenses/agpl-3.0.txt AGPL V3
 * @link      http://jakejohns.net
 */

namespace Jnjxp\Viewd;

use Aura\View\View;

use Psr\Http\Message\RequestInterface as Request;
use Psr\Http\Message\ResponseInterface as Response;

use Aura\Payload_Interface\PayloadInterface as Payload;

/**
 * Abstract Responder
 *
 * @category Responder
 * @package  Jnjxp\Viewd
 * @author   Jake Johns <jake@jakejohns.net>
 * @license  http://www.gnu.org/licenses/agpl-3.0.txt AGPL V3
 * @link     http://jakejohns.net
 */
abstract class AbstractResponder
{
    /**
     * View object
     *
     * @var View
     *
     * @access protected
     */
    protected $view;

    /**
     * Prefix
     *
     * @var null|string
     *
     * @access protected
     */
    protected $prefix;


    /**
     * __construct
     *
     * @param View $view Aura\View object
     *
     * @access public
     */
    public function __construct(View $view)
    {
        $this->view = $view;
    }

    /**
     * Set prefix
     *
     * @param mixed $prefix DESCRIPTION
     *
     * @return mixed
     *
     * @access public
     */
    public function setPrefix($prefix)
    {
        $this->prefix = $prefix;
    }

    /**
     * __invoke
     *
     * @param Request  $request  PSR7 Request
     * @param Response $response PSR7 Response
     * @param Payload  $payload  Domain Payload
     *
     * @return Response
     *
     * @access public
     */
    public function __invoke(
        Request $request,
        Response $response,
        Payload $payload = null
    ) {
        $this->request  = $request;
        $this->response = $response;
        $this->payload  = $payload;

        $this->prepare();
        $this->render();

        return $this->response;
    }

    /**
     * Prepare
     *
     * @return mixed
     *
     * @access protected
     */
    protected function prepare()
    {
        $this->view->addData(['request' => $this->request]);
    }

    /**
     * Render
     *
     * @return mixed
     *
     * @access protected
     */
    protected function render()
    {
        $view = $this->view;
        $this->response->getBody()->write($view());
    }
}


src/Viewd.php
<?php
/**
* Jnjxp\Viewd
*
* PHP version 5
*
* This program is free software: you can redistribute it and/or modify it
* under the terms of the GNU Affero General Public License as published by
* the Free Software Foundation, either version 3 of the License, or (at your
* option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Affero General Public License for more details.
*
* You should have received a copy of the GNU Affero General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*
* @category  Viewd
* @package   Jnjxp\Viewd
* @author    Jake Johns <jake@jakejohns.net>
* @copyright 2015 Jake Johns
* @license   http://www.gnu.org/licenses/agpl-3.0.txt AGPL V3
* @link      http://jakejohns.net
 */

namespace Jnjxp\Viewd;

use Aura\View\View;

use Psr\Http\Message\RequestInterface as Request;
use Psr\Http\Message\ResponseInterface as Response;

/**
 * Viewd
 *
 * @category Viewd
 * @package  Jnjxp\Viewd
 * @author   Jake Johns <jake@jakejohns.net>
 * @license  http://www.gnu.org/licenses/agpl-3.0.txt AGPL V3
 * @link     http://jakejohns.net
 */
class Viewd
{
    /**
     * View object
     *
     * @var View
     *
     * @access protected
     */
    protected $view;

    /**
     * Path prefix for viewd scripts
     *
     * @var string
     *
     * @access protected
     */
    protected $prefix;

    /**
     * Error view name
     *
     * @var string
     *
     * @access protected
     */
    protected $error;

    /**
     * ResponderFactory
     *
     * @var mixed
     *
     * @access protected
     */
    protected $responderFactory;

    /**
     * __construct
     *
     * @param View   $view   view renderer
     * @param string $prefix path prefix for viewd scripts
     * @param string $error  name of error script
     *
     * @access public
     */
    public function __construct(View $view, $prefix = 'viewd', $error = 'error')
    {
        $this->view = $view;
        $this->prefix = $prefix;
        $this->error = $error;
        $this->setResponderFactory([$this, 'responderFactory']);
    }

    /**
     * SetResponderFactory
     *
     * @param callable $factory DESCRIPTION
     *
     * @return mixed
     *
     * @access public
     */
    public function setResponderFactory(callable $factory)
    {
        $this->responderFactory = $factory;
        return $this;
    }

    /**
     * ResponderFactory
     *
     * @return mixed
     *
     * @access protected
     */
    protected function responderFactory()
    {
        return new SimpleResponder($this->view);
    }

    /**
     * __invoke
     *
     * @param Request  $request  server request
     * @param Response $response response
     * @param callable $next     next
     *
     * @return Response
     *
     * @access public
     */
    public function __invoke(
        Request $request,
        Response $response,
        callable $next = null
    ) {
        $name = $this->getName($request);

        if ($this->has($name)) {
            return $this->render($name, $request, $response);
        }

        $newResponse = $next ? $next($request, $response) : $response;

        if ($this->shouldRenderError($request, $newResponse)) {
            return $this->error($request, $newResponse);
        }

        return $newResponse;
    }

    /**
     * ShouldRenderError
     *
     * @param Request  $request  original request
     * @param Response $response response
     *
     * @return mixed
     *
     * @access protected
     */
    protected function shouldRenderError(Request $request, Response $response)
    {
        if ($request->getAttribute('jnjxp/viewd:no-error')) {
            return false;
        }

        if ((string) $response->getBody()) {
            return false;
        }
        return true;
    }

    /**
     * GetName
     *
     * @param Request $request request
     *
     * @return string
     *
     * @access protected
     */
    protected function getName(Request $request)
    {

        if (! $path = $request->getAttribute('jnjxp/viewd:script')) {
            $path = parse_url($request->getRequestTarget(), PHP_URL_PATH);
            $path = trim($path, '/');
        }

        $name = $this->prefix
            . DIRECTORY_SEPARATOR
            . ($path ? $path : 'index');

        return $name;
    }

    /**
     * Has
     *
     * @param string $name name of script
     *
     * @return bool
     *
     * @access protected
     */
    protected function has($name)
    {
        return $this->view->getViewRegistry()->has($name);
    }

    /**
     * Render
     *
     * @param string   $name     name of script to render
     * @param Request  $request  request
     * @param Response $response response
     *
     * @return Response
     *
     * @access protected
     */
    protected function render($name, Request $request, Response $response)
    {
        $responder = call_user_func($this->responderFactory);
        $request = $request->withAttribute('view', $name);
        return $responder($request, $response);
    }

    /**
     * Error
     *
     * @param Request  $request  request
     * @param Response $response response
     *
     * @return Response
     *
     * @access protected
     */
    protected function error(Request $request, Response $response)
    {
        return $this->render(
            $this->error,
            $request,
            $response->withStatus(404)
        );
    }
}
